lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +763,37 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts given `date` by `duration` according to its calendar.
+
+  Available units are: `:year, :month, :week, :day`.
+
+  Raises ArgumentError when called with time units.
+
+  ## Examples
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      {:ok, ~D[2016-03-03]}
+      iex> Date.shift(~D[2016-01-30], month: 1)
+      {:ok, ~D[2016-02-29]}
+      iex> Date.shift(~D[2016-01-31], year: 4, day: 1)
+      {:ok, ~D[2020-02-01]}
+      iex> Date.shift(~D[2016-01-03], Duration.new(month: 2))
+      {:ok, ~D[2016-03-03]}
+
+  """
+  @doc since: "1.7.0"
+  @spec shift(Calendar.date(), Duration.t() | [Duration.unit()]) :: {:ok, t}
+  def shift(%Date{calendar: calendar} = date, %Duration{} = duration) do
```

Comment:
Btw, you should implement this like we do `add`. We don't pattern match on the `Date` struct, only on the fields. This allows downcasting. Almost all calendar APIs work like this (we need to update the other shift functions accordingly).